### PR TITLE
 UCLA miniscope V4 support

### LINKED ISSUE #112: Miniscope V4 support
Add support for miniscope 4.0

### LINKED ISSUE #133: Miniscope real-time properties need to be moved to configuration node
- See #114

### PR DESCRIPTION
**Pull request by jonnew**
- Tested on OEPS miniscope V4.4
- To finalize this, we must test on a large batch of scopes to ensure functionality of all features and reliability of datastream
- Fixes #112 
- Fixes #133

### CODE REVIEW

File: OpenEphys.Onix1/UclaMiniscopeV4CameraData.cs [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+                    var clockBuffer = new ulong[UclaMiniscopeV4.SensorRows];
+                    var awaitingFrameStart = true;
+
+                    var frameObserver = Observer.Create<oni.Frame>(
+                        frame =>
+                        {
+                            var payload = (UclaMiniscopeV4ImagerPayload*)frame.Data.ToPointer();
+
+                            // Wait for first row
+                            if (awaitingFrameStart && (payload->ImageRow[0] & 0x8000) == 0)
```
Comment (jonnew) [inline OpenEphys.Onix1/UclaMiniscopeV4CameraData.cs:51]: Replace `awaitingFrameStart` with `sampleIndex == 0`

**Comment by jonnew** [comment]
@bparks13 Please have a look at the latest commits if you dont mind, this is ready to merge on my end.

File: OpenEphys.Onix1/ConfigureUclaMiniscopeV4.cs [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+using System.Threading;
+using oni;
+
+namespace OpenEphys.Onix1
+{
+    /// <summary>
+    /// Configures a UCLA Miniscope V4 on the specified port.
+    /// </summary>
+    /// <remarks>
+    /// The UCLA Miniscope V4 is a e miniaturized fluorescent microscope for performing single-photon calcium
```
Comment (bparks13) [inline OpenEphys.Onix1/ConfigureUclaMiniscopeV4.cs:13]: "Miniscope V4 is a _e_ miniaturized" => "Miniscope V4 is a miniaturized"

File: OpenEphys.Onix1/ConfigureUclaMiniscopeV4.cs [hunk 3]
```diff
@@ -0,0 +1,198 @@
...
+    /// <item><description>A Python-480 0.48 Megapixel CMOS image sensor.</description></item>
+    /// <item><description>A BNO055 9-axis IMU for real-time, 3D orientation tracking.</description></item>
+    /// <item><description>An electrowetting lens for remote focal plane adjustment.</description></item>
+    /// <item><description>An excitation LED with adjustable brightness control and optional exposure-driven
+    /// interleaving to reduce photobleaching.</description></item>
+    /// </list>
+    /// </remarks>
+    public class ConfigureUclaMiniscopeV4 : MultiDeviceFactory
+    {
+
```
Comment (bparks13) [inline OpenEphys.Onix1/ConfigureUclaMiniscopeV4.cs:25]: nit: extra line

File: OpenEphys.Onix1/ConfigureUclaMiniscopeV4.cs [hunk 4]
```diff
@@ -0,0 +1,198 @@
...
+        /// port voltage.
+        /// </para>
+        /// <para>
+        /// Warning: this device requires 4.0 to 5.0V, measured at the miniscope, for proper operation. Supplying higher
+        /// voltages may result in damage.
+        /// </para>
+        /// </remarks>
+        [Description("If defined, it will override automated voltage discovery and apply the specified voltage" +
+                     "to the miniscope. Warning: this device requires 4.0 to 5.0V, measured at the scope, for proper operation." +
+                     "Supplying higher voltages may result in damage to the miniscope.")]
```
Comment (bparks13) [inline OpenEphys.Onix1/ConfigureUclaMiniscopeV4.cs:99]: Need spaces after "specified voltage" and "proper operation.". When these are rendered in Bonsai the resulting string is "voltageto" and "operation.Supplying".

File: OpenEphys.Onix1/UclaMiniscopeV4Bno055Data.cs [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+        [Category(DeviceFactory.ConfigurationCategory)]
+        public string DeviceName { get; set; }
+
+        /// <summary>
+        /// Generates a sequence of <see cref="Bno055DataFrame"/>s at approximately 100 Hz.
+        /// </summary>
+        /// <returns>A sequence of <see cref="Bno055DataFrame"/>s.</returns>
+        /// <remarks>
+        /// This will generate a sequence of <see cref="Bno055DataFrame"/>s at approximately 100 Hz. This rate
+        /// may be limited by the hardware I2C bus.
```
Comment (bparks13) [inline OpenEphys.Onix1/UclaMiniscopeV4Bno055Data.cs:27]: I believe we have removed mention of the I2C bus in the other BNO data nodes, should we remove this here too?

File: OpenEphys.Onix1/UclaMiniscopeV4Bno055Data.cs [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+        /// <summary>
+        /// Generates a sequence of <see cref="Bno055DataFrame"/>s.
+        /// </summary>
+        /// <param name="source">An input sequence that drives the production of <see
+        /// cref="Bno055DataFrame"/>s</param>
+        /// <returns>A sequence of <see cref="Bno055DataFrame"/>s.</returns>
+        /// <remarks>
+        /// A <see cref="Bno055DataFrame"/> will be produced each time an element is received from the
+        /// <paramref name="source"/> sequence. This rate is limited by the hardware I2C bus and has a
+        /// maximum of 100 Hz.
```
Comment (bparks13) [inline OpenEphys.Onix1/UclaMiniscopeV4Bno055Data.cs:45]: Is the 100 Hz a hard limit imposed by the hardware? For this overload, does that same limit exist, or is this overload simply polling data at whatever interval data is requested (within reason)?

**Review by bparks13 — APPROVED** [review]
Looks good, couple minor typos but that's it.

File: OpenEphys.Onix1/UclaMiniscopeV4Bno055Data.cs (hunk 6 above)
Comment (jonnew) [reply OpenEphys.Onix1/UclaMiniscopeV4Bno055Data.cs]: Yes it is. the Bno itself can only updated the orientation estimate at 100 Hz, so polling faster will result in repeated data.

**Comment by jonnew** [comment]
@bparks13 I had to add some extra stuff, in there, and XML comments for that. Can you have a look at the new comments for typos etc? Thank you.

**Review by bparks13 — APPROVED** [review]

**Pull request merged**